### DISCUSSION TITLE
Calibration in the slicer is a great idea - lets make it easier

### DISCUSSION THREAD

**Discussion by julie777** [OrcaSlicer/OrcaSlicer#701] [Ideas]
Being able to take a new filament, enter in the manufacturers numbers and then run tests to dial it in is a great thing, especially when it is mostly automated. I think we should take it a step further.

Make calibration work like a wizard. 
1. select calibration
2. a dialog opens that shows a list of tests. Select the tests you want to run.
3. select the printer profile and the filament to use.
```
I have question about the current documentation when it talks about selecting the process, because it changes many of those settings for a given test. Is the choice of .20 standard vs. .20 strength really part of the test?
```
4. select run tests
5. the calibration program creates a copy of the selected filament, and other settings if necessary to create a test result configuration. This will be modified during the testing by the calibration wizard. After all, it is easier for the calibration code to do arithmetic that for use to do it.
6. calibration prints some objects for a test
7. calibration explains what to look for and then waits for your answer
8. calibration uses your answer to make adjustments to the settings/config.
9. go back to step 6 for each plate that needs to be printed.
10. when all tests have been run the calibration wizard asks you to name the newly produced configuration.

If done in a way that reduces printing time as much as possible, and has information provided about what each test is for so you only select the tests you need, this could be a process that is very efficient. 

It could be something you do when you get a new filament brand, type, or even color.
If done well, fine tuning a filament based on an already tested filament might be very fast. For example, you have brand X filament in several colors that are working well. You use a new color and the results are not as good as you would like, so you start with the existing tuned config for that brand and run calibration for the new color, which should be relatively close to what you already have.

The key to this is automate as much of the process as possible, do every step that can be done automatically and only ask the user to make decisions about print quality, and make those decisions easier (related post #702). During the current process there is to much "new project", "do you want to save settings", "what order should I run the tests", etc.

**Comment by julie777** [comment]
I have been calibrating some new filament and have even more thoughts about this.

Imagine that calibration was a wizard. 

1.  You click calibrate and the wizard prepares for calibration
    1. you are prompted to save or discard the current project
    1. the wizard opens, presenting a list of the various calibration choices.
    1.  You choose the tests that you would like to run.
    1.  you choose the printer, and the preset for printer settings, filament, and process, as you would for any print job. Those settings form the foundation of the calibration.
    1.  Note: It was not obvious to me originally that some of these choices are the basis for calibration
2.  you start start the calibration procedure
3.  The wizard chooses the calibration to perform
    1.  the wizard configures the test model, slices it, and sends it to the printer
    1.  when the print is complete, the wizard notifies you (notification, sound, etc.) so that you know the print has finished and you need to retrieve the objects from the printer for analysis
    1.  the wizard opens a new page describing how to analyze the result 
    1.  the wizard prompts you for the information it needs
    1.  the wizard does any calculations necessary using the information that you provided
    1.  the wizard makes any setting adjustments necessary (such as changing filament temp, or flow ratio
    1. the wizard repeats step 3 for each remaining test.
6. the wizard gives you the option to print a test print using all the settings determined by the test. This could be a benchy, or my preference is this model https://www.printables.com/model/395946-the-totally-mundane-lit-calibration-cube. In any case, there can be several choices presented for the user to choose from.
7. When the final test print completes, or if the user opts to skip it, the wizard leaves the slicer set with a new project created and the originally chosen presets with settings changed based on the calibration tests, so they can be saved and named them as desired.


This has many benefits
- the wizard chooses the best order to run the selected tests
- the wizard can provide an estimate of the total time required after the set of tests is selected.
- the wizard can run modified machine start gcode to speed up the testing and prevent things like having the start code perform flow testing on the X1
- there are no confirmations to answer about saving settings or the like during the wizard operation
- there are no calculations for the user to perform
- the user can select any group of tests that they want to perform
- the tests can include any combination of tests for printer settings, filament settings, or process settings
- the user does not have to monitor the printing process because the wizard will provide notification when the user is required to analyze something. This makes sure the calibration can run as quickly as possible
- while the wizard is running, 
  - it takes over the slicer and runs in a modal state. 
  - during printing the wizard will show the device page with the camera on if available
  - the user can abort the current print if there is a problem and the wizard will provide the option to redo the print, or cancel the entire calibration proceedure
  - this ensures the integrity of the calibration

**Comment by julie777** [comment]
One example of how calibration is harder than it should be:
I finished writing the above and realized that my print was complete. I was printing the PA tower, but I had set custom values in the dialog. I don't remember what I set them to, and when I reopened the PA dialog it is showing the defaults.

**Comment by jjl40** [comment]
this is my first time using Orca slicer. Trying to figure out what the flow calibration was very confusing. 
First: how do you find what the flow ratio is? I did not realize that this was shown at the very end. of the description. Should be first.
Second: the instructions say "** Remember** to save the filament profile.". As a first time user, I'm wondering, how?
Third: the screen animation is really confusing. It's too fast. When does a step begin and end. A speed control and/or a pause button would be a big help.
I think after failing the first time I might be able to do it correctly now. If I can figure out the how to ** Remember** to save the filament profile.

**Comment by Edzi11** [comment]
I know this is was posted a while back but I truly like this concept. As the 3D printing world is moving to being more automated this is definitely something that should be implemented into orca. The newest version of the official Bambu Studio is already doing something similar to this when manually calibrating flow and PA, I believe orca could do this even better. This slicer has already become my one stop shop for all my printers so seeing this implemented would be amazing. Maybe submit a feature request if you already havent.

**Comment by jjl40** [comment]
After using the calibration processes for a while, I have developed a love/hate attitude towards it. Absolutely a great idea and very helpful bringing the critical functions of a 3D printer into focus. However, it definitely needs some polish.
!. It is too terse for even some experienced users. 
2. (example) The flow rate cal has several errors/omissions. The directions say to select the one that has "the smoothest top surface." The annotations in the picture say to observe the edges for signs of under or over extrusion. Honestly, the pictures are bad to the point of distraction. Even when printed on my old Anycubic Mega Pro, all the blocks look better than any in the pictures.
This is a very useful tool as is but it's so frustrating because it forces one to think how good it could be. Especially for someone new to the hobby or, like me, trying to systematically dial in a new printer.
Orca has come a long way in a short amount of time. With new users jumping on board and an avalanche of new products hitting the market, it's obvious that your developer's plates are more than full. Improving a bunch of scripts to improve a calibration tool is not the most glamorous coding challenge but just look at how many calibration tools are on Thingiverse and it is evident that there is a huge market crying out for it.
Keep up the good work!
